src/IR.h

```diff
@@ -886,16 +886,44 @@ struct Atomic : public StmtNode<Atomic> {
     static const IRNodeType _node_type = IRNodeType::Atomic;
 };
 
-/** Represent a target-specific vector instruction.
- *  Intrinsic may not be element-wise operation, i.e.
- *  dot_products. */
-struct VectorIntrinsic : public ExprNode<VectorIntrinsic> {
-    std::string name;
+/** Represent a length-agnostic and target-specific
```

Comment:
I gather the difference from the existing approach (```Call``` node) is that the result type and the input types do not have to have the same shape. Is this the defining characteristic of when one would use this vs. ```Call```?

The use of "target specific" needs to be clarified. E.g. some of the options are not target specific. Moreover, if the operation differs in some way from the exact definition of the machine instruction names, such as providing normalization of length, the differences from the machine instruction need to be documented.

LLVM is adding the vector predicated instructions and my guess is this is going to become the way to generate LLVM vector IR. These take predication masks and a vector length. One possibility is to add those parameters as field of this node.

Finally, how this node interacts with various code generation options needs to be documented. Specifically, I would expect actually target specific instructions -- e.g. those documented as for x86 -- to only be allowed when targeting the specific target. It may or may not be the case that it has to be via LLVM, but is asking for an x86 instruction on another architecture supposed to work? (I.e. are we planning to provide target independent lowering for all of these or only some of them?)

Reply:
Thanks for the feedback! These are excellent questions/comments.

Yes, the main difference between `Call` nodes and this node is that `Call` nodes are element-wise operations, so they cannot express operations with a different output vector length than the inputs.

The intent behind "target specific" is that these nodes should not be created for backends that do not implement the operation. However, there are some operations that may exist on multiple backends (i.e. `abs`), that are expected to have the same semantics.

> One possibility is to add those parameters as field of this node.

That could absolutely be a possibility for the future, however this node is not intended to be used for every single vector operation (i.e. we see no reason to use it for a simple vector addition), only for target-specific instructions.

> Specifically, I would expect actually target specific instructions -- e.g. those documented as for x86 -- to only be allowed when targeting the specific target.

Yes, that is required for these nodes. I have yet to come up with a good abstraction for enforcing this though, especially because many of the x86-specific instructions actually require SSE41 (i.e. the predicated patterns in X86Optimize.cpp ensure that the target has the required feature before generating a `VectorInstruction` node).

> It may or may not be the case that it has to be via LLVM, but is asking for an x86 instruction on another architecture supposed to work? (I.e. are we planning to provide target independent lowering for all of these or only some of them?)

No, we absolutely do not intend to implement lowering for instructions on other architectures. x86 instruction nodes should only be generated when the target is x86 and that instruction is supported on the target. This is very different from `Call` intrinsics, which must all be supported on any architecture.
For example, x86 does not support `absd`, and patterns in X86Optimize.cpp contains patterns for efficiently lowering this instruction onto x86.